Homework 5_4

ZubenkoHalyna opened the pull request.

ZubenkoHalyna committed:
implementation added

ZubenkoHalyna committed:
program divided on tree files

ZubenkoHalyna committed:
static field fixed

ZubenkoHalyna committed:
ConsoleHelper made stateless

dmytro-hryhoruk commented on src/main/java/core/basesyntax/FileHelper.java:
```diff
@@ -0,0 +1,149 @@
+import java.time.format.DateTimeFormatter;
+import java.util.HashMap;
+
+public class FileHelper {
```
that's a lot constants. Let's remove those that you are using as exceptions message, param_titles(those numbers could be just local variables)

dmytro-hryhoruk commented on src/main/java/core/basesyntax/FileHelper.java:
```diff
@@ -0,0 +1,149 @@
+        if (instance == null) {
+            instance = new FileHelper();
+        }
+        return instance;
```
it's better to have separate class for singleton https://habr.com/ru/post/27108/

dmytro-hryhoruk commented on src/main/java/core/basesyntax/FileHelper.java:
```diff
@@ -0,0 +1,149 @@
+                SIZE_PARAM_TITLE, file.length(),
+                NUMBER_OF_STRINGS_PARAM_TITLE, numberOfStrings,
+                NUMBER_OF_WORDS_PARAM_TITLE, numberOfWords,
+                NUMBER_OF_CHARS_PARAM_TITLE, numberOfChars);
```
we don't need these string to be constants)

dmytro-hryhoruk commented on src/main/java/core/basesyntax/FileHelper.java:
```diff
@@ -0,0 +1,149 @@
+                NUMBER_OF_CHARS_PARAM_TITLE, numberOfChars);
+    }
+
+    void showFileContent(String path, String fileName) {
```
access modifiers?

dmytro-hryhoruk commented on src/main/java/core/basesyntax/MainApp.java:
```diff
@@ -0,0 +1,85 @@
+    private static final String EXIT_COMMAND = "EXIT";
+
+    public static void main(String[] args) {
+        for (; ; ) {
```
move all logic, constants  to separate class.And call it's methods from main  method

dmytro-hryhoruk commented on src/main/java/core/basesyntax/MainApp.java:
```diff
@@ -0,0 +1,85 @@
+    private static final String WRONG_NUMBER_OF_ARGS_ERROR_MSG =
+            "Wrong number of arguments for %s command.\n";
+    private static final String REG_EXP_MULTI_WHITESPACE = "[\\s]+";
+    private static final String CREATE_COMMAND = "CREATE";
+    private static final String READ_COMMAND = "READ";
+    private static final String INFO_COMMAND = "INFO";
+    private static final String HELP_COMMAND = "HELP";
+    private static final String EXIT_COMMAND = "EXIT";
```
we already have similar enum don't we ?

dmytro-hryhoruk commented on src/main/java/core/basesyntax/FileHelper.java:
```diff
@@ -0,0 +1,149 @@
+        int numberOfStrings = 0;
+        int numberOfWords = 0;
+        int numberOfChars = 0;
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String str;
+            while ((str = reader.readLine()) != null) {
+                numberOfStrings++;
+                numberOfChars += str.length();
+                numberOfWords += str.split(REG_EXP_MULTI_WHITESPACE).length;
```
You can make your method much shorter if use nio Files.readAllLines(file).size() = lines.And you can put it right into your sout. Google method for other cases)

dmytro-hryhoruk commented on src/main/java/core/basesyntax/ConsoleHelper.java:
```diff
@@ -0,0 +1,37 @@
+    private static final String CORRECT_YES_ANSWER = "y";
+    private static ConsoleHelper instance;
+
+    public static ConsoleHelper getInstance() {
+        if (instance == null) {
+            instance = new ConsoleHelper();
+        }
+        return instance;
+    }
```
create separate Factory class which will give instance of ConsoleHelper and FileHelper

ZubenkoHalyna replied:
I try to implement Factory pattern for creating helpers and use it in tests for isolation MainManu class from console input and file output. So ConsoleHelper become a stub for input, and FileHelper insted of doing operations with files just checks if a correct method of FileHelper was invoked.
But I can't generate more informative names for Factories than RealFactory in main app and MockFactory in tests. In main app it looks confusing as for me, but I don't know what to do with it.

ZubenkoHalyna committed:
Number of constants reduced
HelperFactory added
Tests added
logic from main method moved to MaimMenu class
NIO used instead of IO
Singleton implemented with static class usage

ZubenkoHalyna committed:
Refactoring. Provided test cases independence
